lib/ansible/plugins/httpapi/fortios.py

```diff
@@ -0,0 +1,134 @@
+author:
+    - Miguel Angel Muñoz (@magonzalez)
+httpapi : fortios
+short_description: HttpApi Plugin for Fortinet FortiOS Appliance or VM
+description:
+  - This HttpApi plugin provides methods to connect to Fortinet FortiOS Appliance or VM via REST API
+version_added: "2.9"
+"""
+
+from ansible.plugins.httpapi import HttpApiBase
+from ansible.module_utils.basic import to_text
+import urllib
+import json
+import re
+
+
+class HttpApi(HttpApiBase):
+    def __init__(self, connection):
+        super(HttpApi, self).__init__(connection)
+
+        self._ccsrftoken = ''
+
+    def set_become(self, become_context):
+        """
+        Elevation is not required on Fortinet devices - Skipped
+        :param become_context: Unused input.
+        :return: None
+        """
+        return None
+
+    def login(self, username, password):
+        """Call a defined login endpoint to receive an authentication token."""
+
+        data = "username=" + urllib.parse.quote(username) + "&secretkey=" + urllib.parse.quote(password) + "&ajax=1"
+        dummy, result_data = self.send_request(url='/logincheck', data=data, method='POST')
+        if result_data[0] != '1':
+            raise Exception('Wrong credentials. Please check')
+
+    def logout(self):
+        """ Call to implement session logout."""
+
+        self.send_request(url='/logout', method="POST")
+
+    def update_auth(self, response, response_text):
+        """
+        Get cookies and obtain value for csrftoken that will be used on next requests
+        :param response: Response given by the server.
+        :param response_text Unused_input.
+        :return: Dictionary containing headers
+        """
+
+        headers = {}
+
+        for attr, val in response.getheaders():
+            if attr == 'Set-Cookie' and 'APSCOOKIE_' in val:
+                headers['Cookie'] = val
+
+            elif attr == 'Set-Cookie' and 'ccsrftoken=' in val:
+                csrftoken_search = re.search('\"(.*)\"', val)
+                if csrftoken_search:
+                    self._ccsrftoken = csrftoken_search.group(1)
+                    headers['x-csrftoken'] = self._ccsrftoken
+
+        return headers
+
+    def handle_httperror(self, exc):
+        """
+        Not required on Fortinet devices - Skipped
+        :param exc: Unused input.
+        :return: exc
+        """
+        return exc
+
+    def send_request(self, **message_kwargs):
+        """
+        Responsible for actual sending of data to the connection httpapi base plugin.
+        :param message_kwargs: A formatted dictionary containing request info: url, data, method
+
+        :return: Status code and response data.
+        """
+        url = message_kwargs.get('url', '/')
+        data = message_kwargs.get('data', '')
+        method = message_kwargs.get('method', 'GET')
+
+        if self._ccsrftoken == '' and not (method == 'POST' and 'logincheck' in url):
+            raise Exception('Not logged in. Please login first')
```

Comment:
Perhaps I'm missing something, but I just don't see how this check should ever trigger, nor why the user should get this message.

* Your `login()` method gets called automatically as the first request to the device. The response from that will pass through `update_auth()` and set `'x-csrftoken'`
* Requests are made to the API. Because you returned the csrf token in `update_auth()` previously, that is automatically added to the request headers without having to add them below.
* At some point I guess you clear the token? This is the part I don't understand. Furthermore, as you indicate that `handle_httperror()` is unneeded, I'm presuming that the API returns its errors in 200 responses for some reason.
* So you can parse the access denied error from the response to `send()` below, and if you need to login again, you call `login()` again before retrying the call and failing if it doesn't work a second time. Or you could let it pass all the way back to the module_utils, reset the connection and retry, depending on where this sort of thing is likely to happen.

Do I have something wrong here? The whole point of httpapi is to abstract the authentication dance _away_ from the user, otherwise we might as well go back to calling `open_url()` in module_utils. I'm just struggling to see why telling the user they have to login is ever a thing that should happen.

Reply:
Hi Qalthos,
 thanks for your answer. If the system guarantees that the first call is a login then I think you are right and the exception will never raise. I can remove that part, thanks for the clarification. I just added it as an extra 'safety' measure.

Regarding the csrf token lost, our device sends the csrf token and the cookie in the first login, but in the later request it may send only the cookie, or nothing. There is no rule about that. However whenever we send a PUT or POST we must add both the token and the cookie always. And here is where I found that if I don't store the csrf token it will be lost in later requests. I have run some tests, and the ansible system do not keep the csrftoken after the first request (even if I return "None" in update_auth). Therefore, I see no other way to make it work than storing the csrf token in the class. Please let me know if I am doing something wrong. This is the code I am testing according to your comments but it does not work bcs later requests do not keep the csrf token:
```
    def update_auth(self, response, response_text):

        headers = {}

        for attr, val in response.getheaders():
            if attr == 'Set-Cookie' and 'APSCOOKIE_' in val:
                headers['Cookie'] = val

            elif attr == 'Set-Cookie' and 'ccsrftoken=' in val:
                csrftoken_search = re.search('\"(.*)\"', val)
                if csrftoken_search:
                    headers['x-csrftoken'] = self._ccsrftoken

        if len (headers) == 2:
            return headers
        else:
            return None

    def send_request(self, **message_kwargs):

        url = message_kwargs.get('url', '/')
        data = message_kwargs.get('data', '')
        method = message_kwargs.get('method', 'GET')

        try:
            response, response_data = self.connection.send(url, data, #headers=headers,
                                                           method=method)

            return response.status, to_text(response_data.getvalue())
        except Exception as err:
            raise Exception(err)
```
Please let me know if I am doing something wrong or if I misunderstood something.
Thanks!

Reply:
Hi Qalthos,
 thanks for your answer. If the system guarantees that the first call is a login then I think you are right and the exception will never raise. I can remove that part, thanks for the clarification. I just added it as an extra 'safety' measure.

Regarding the csrf token lost, our device sends the csrf token and the cookie in the first login, but in the later request it may send only the cookie, or nothing. There is no rule about that. However whenever we send a PUT or POST we must add both the token and the cookie always. And here is where I found that if I don't store the csrf token it will be lost in later requests. I have run some tests, and the ansible system do not keep the csrftoken after the first request (even if I return "None" in update_auth). Therefore, I see no other way to make it work than storing the csrf token in the class. Please let me know if I am doing something wrong. This is the code I am testing according to your comments but it does not work bcs later requests do not keep the csrf token:
```
    def update_auth(self, response, response_text):

        headers = {}

        for attr, val in response.getheaders():
            if attr == 'Set-Cookie' and 'APSCOOKIE_' in val:
                headers['Cookie'] = val

            elif attr == 'Set-Cookie' and 'ccsrftoken=' in val:
                csrftoken_search = re.search('\"(.*)\"', val)
                if csrftoken_search:
                    headers['x-csrftoken'] = self._ccsrftoken

        if len (headers) == 2:
            return headers
        else:
            return None

    def send_request(self, **message_kwargs):

        url = message_kwargs.get('url', '/')
        data = message_kwargs.get('data', '')
        method = message_kwargs.get('method', 'GET')

        try:
            response, response_data = self.connection.send(url, data, #headers=headers,
                                                           method=method)

            return response.status, to_text(response_data.getvalue())
        except Exception as err:
            raise Exception(err)
```
Please let me know if I am doing something wrong or if I misunderstood something.
Thanks!

Reply:
> I have run some tests, and the ansible system do not keep the csrftoken after the first request (even if I return "None" in update_auth).

This is what I was alluding to in my previous comments- if you need to keep the token for that purpose, then I have no problem with that. My main goal was to impress that you are doing a bunch of things you don't have to inside of `send_request()`.

Multiple required tokens is a thing I admit I hadn't considered before. It's possible that instead of overwriting `_auth` in the connection, we should instead update the dictionary. I'm not immediately sure if there are any problems that would cause, but that would also fix your issue, I think.

Reply:
Ok, thanks for the clarification @Qalthos.
I have added a commit with the changes, hoping to reflect your comments as much as possible. I have reduced to a minimum the things done in 'send_request' and now I return always the csrftoken as headers in update_auth. Do you think it could be ok this way?